_drafts/cortex-m-breakpoints.md

```diff
@@ -0,0 +1,607 @@
+
+> A full discussion of debug interfaces is outside the scope of this article but more details can be found in [our post on the topic]({% post_url 2019-08-06-a-deep-dive-into-arm-cortex-m-debug-interfaces %}). We'll recap the parts relevant for this post below.
+
+At a high level, `gdb` (the "client") interfaces with the embedded MCU via a `gdbserver` (in our case SEGGER's JLinkGDBServer). The protocol talked between the gdb "client" and the gdb "server" is referred to as "GDB Remote Serial Protocol" (**GDB RSP**)[^4].
+
+If we take a look at the GDB RSP docs we will find that the ['z' packets](https://sourceware.org/gdb/onlinedocs/gdb/Packets.html#index-z-packet) are used to configure breakpoints.
+
+#### GDB RSP Breakpoint Command
+
+The format of the commands are `‘Z/z type,addr,kind’` where:
+
+| GDB Remote Serial Packet | Operation                            |
+| ------------------------ | ------------------------------------ |
+| `Z0,addr,kind`           | Add Software Breakpoint at `addr`    |
+| `z0,addr,kind`           | Remove Software Breakpoint at `addr` |
+| `Z1,addr,kind`           | Add Hardware Breakpoint at `addr`    |
+| `z1,addr,kind`           | Remove Hardware Breakpoint at `addr` |
+
+`kind` is architecture specific but for ARM it indicates what instruction to use. (The instruction changes depending on whether the MCU uses the ARM or Thumb instruction set).[^9]
+
+> NOTE: The use of "software" breakpoint in GDB RSP is a bit of a misnomer. The gdb server can choose how to best install the breakpoint. For example, the server may opt to install a "software" breakpoint as a hardware breakpoint if there are still ones available.
+
+### Debugging GDB Remote Serial Protocol
+
+With that out of the way, let's take a look at the actual **GDB RSP** commands that get sent over the wire when we install a breakpoint. Conveniently GDB has a builtin debug command, `set debug remote 1`, which can be used to enable printing of every single command sent between the client and server!
+
+To facilitate breakpoint testing, I've created some test functions in the demo app `dummy_function_*` that we can install breakpoints on. We can use the `dump_dummy_funcs` CLI command to print information about them all:
+
+```
+shell> dump_dummy_funcs
+dummy_function_1: Starts at 0x240. First Instruction = 0x48024901
+dummy_function_2: Starts at 0x250. First Instruction = 0x48024901
+dummy_function_3: Starts at 0x260. First Instruction = 0x48024901
+dummy_function_4: Starts at 0x270. First Instruction = 0x48024901
+dummy_function_5: Starts at 0x280. First Instruction = 0x48024901
+dummy_function_6: Starts at 0x290. First Instruction = 0x48024901
+dummy_function_7: Starts at 0x2a0. First Instruction = 0x48024901
+dummy_function_8: Starts at 0x2b0. First Instruction = 0x48024901
+dummy_function_9: Starts at 0x2c0. First Instruction = 0x48024901
+dummy_function_ram: Starts at 0x20000000. First Instruction = 0x48024901
+```
+
```

Comment:
Add a sentence about why you dump the address & first instruction, and maybe a code snippet (or link to code) that shows these are no-ops? In case someone wants to repro this.